File: legacy/web/openexchangerates.py
# Standard library imports
import requests
import json

# Local imports
from web import config


class OpenExchangeRatesException(Exception):
	pass


def get() -> dict:
	params = {'app_id': config.OPENEXCHANGERATES_APPID, 'base': 'USD'}
	response = requests.get(
		'https://openexchangerates.org/api/latest.json',
		params=params
	)
	response.raise_for_status()
	response = json.loads(response.text)
	return response['rates']

File: src/lib/exchange-rates/openexchangerates.ts
const OPENEXCHANGERATES_URL = 'https://openexchangerates.org/api/latest.json';

type LatestRatesResponse = {
  rates: Record<string, number>;
};

export async function fetchLatestExchangeRates(): Promise<Record<string, number>> {
  const appId = process.env.OPENEXCHANGERATES_APPID;
  if (!appId) throw new Error('OPENEXCHANGERATES_APPID is not set');

  const url = new URL(OPENEXCHANGERATES_URL);
  url.searchParams.set('app_id', appId);
  url.searchParams.set('base', 'USD');

  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Open Exchange Rates request failed (${response.status})`);
  }

  const body = (await response.json()) as LatestRatesResponse;
  if (!body.rates || typeof body.rates !== 'object') {
    throw new Error('Open Exchange Rates response missing rates');
  }

  return body.rates;
}
